### PR TITLE
"Handle" calls to upstream monomorphizations in compiler_builtins

### LINKED ISSUE #121552: Debug builds of compiler_builtins fail to link due to references to precondition_check in libcore
Likely caused by #120594. cc @saethlin 

CI for compiler_builtins is failing with:
```
error: linking with `link.exe` failed: exit code: 1120
  |
  = note: "C:\\Program Files\\Microsoft Visual Studio\\2022\\Enterprise\\VC\\Tools\\MSVC\\14.38.33130\\bin\\HostX64\\x64\\link.exe" "/NOLOGO" "C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\rustccNjGuE\\symbols.o" "D:\\a\\compiler-builtins\\compiler-builtins\\target\\x86_64-pc-windows-msvc\\debug\\examples\\intrinsics.intrinsics.94b4869361004a03-cgu.0.rcgu.o" "/LIBPATH:D:\\a\\compiler-builtins\\compiler-builtins\\target\\x86_64-pc-windows-msvc\\debug\\deps" "/LIBPATH:D:\\a\\compiler-builtins\\compiler-builtins\\target\\debug\\deps" "/LIBPATH:C:\\Users\\runneradmin\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib" "kernel32.lib" "msvcrt.lib" "D:\\a\\compiler-builtins\\compiler-builtins\\target\\x86_64-pc-windows-msvc\\debug\\deps\\libcompiler_builtins-b84df12ffea9ddc5.rlib" "/NXCOMPAT" "/LIBPATH:C:\\Users\\runneradmin\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib" "/OUT:D:\\a\\compiler-builtins\\compiler-builtins\\target\\x86_64-pc-windows-msvc\\debug\\examples\\intrinsics.exe" "/OPT:REF,NOICF" "/DEBUG" "/NATVIS:C:\\Users\\runneradmin\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\etc\\intrinsic.natvis" "/NATVIS:C:\\Users\\runneradmin\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\etc\\liballoc.natvis" "/NATVIS:C:\\Users\\runneradmin\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\etc\\libcore.natvis" "/NATVIS:C:\\Users\\runneradmin\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\etc\\libstd.natvis"
  = note: libcompiler_builtins-b84df12ffea9ddc5.rlib(compiler_builtins-b84df12ffea9ddc5.compiler_builtins.ca00aaabc6ad635b-cgu.2.rcgu.o) : error LNK2019: unresolved external symbol _ZN4core3ptr13read_volatile18precondition_check17ha1e11d6217b519e2E referenced in function _ZN4core3ptr13read_volatile17h3fd78456f42a46ecE
          libcompiler_builtins-b84df12ffea9ddc5.rlib(compiler_builtins-b84df12ffea9ddc5.compiler_builtins.ca00aaabc6ad635b-cgu.2.rcgu.o) : error LNK2019: unresolved external symbol _ZN4core10intrinsics19copy_nonoverlapping18precondition_check17h5ef2e74b4a074627E referenced in function _ZN4core3ptr14read_unaligned17h327557c7a8527effE
          D:\a\compiler-builtins\compiler-builtins\target\x86_64-pc-windows-msvc\debug\examples\intrinsics.exe : fatal error LNK1120: 2 unresolved externals
```

The cause seems to be from a combination of factors:
- compiler-builtins cannot reference any symbols in libcore. It can only call functions marked with `#[inline]`, since those are codegen'ed into compiler_builtins.rlib.
- `precondition_check` is marked as `#[inline(never)]`, which breaks this.
- The call is not eliminated because `core::intrinsics::debug_assertions` is resolved late in the codegen backend, after any MIR optimizations.
- This failure happens in debug builds, so LLVM isn't eliminating the `if` branch either.

### LINKED ISSUE #122702: Add codegen/linkage tests for building compiler_builtins with debug assertions
This is similar to https://github.com/rust-lang/rust/issues/118393. If someone figures out how to do one it should be easy to do the other as well.

For a long time, panics from `compiler_builtins` would result in linker errors unless they happened to be optimized out by the linker before resolution. Since https://github.com/rust-lang/rust/pull/122580, such calls will either be replaced by a backend's abort intrinsic or a compile error during code generation.

I've used this locally to assess whether the PR fixes the root problem:
```
RUSTFLAGS="-Copt-level=1 -Cdebug-assertions=yes" cargo +nightly build -Zbuild-std --target=x86_64-unknown-linux-gnu
objdump -r target/x86_64-unknown-linux-gnu/debug/deps/libcompiler_builtins-b10ebdb7a18763a9.rlib | grep core
```
This `objdump -r | grep` is searching for mentions of `core` in the relocations inside the rlib. If don't find any, then `compiler_builtins` does not depend on `core` at link-time. There may be a more fine-tuned way to write this check that would make a test more robust, but this grep is sufficient to validate the property we need.

### PR DESCRIPTION
**Pull request by saethlin**
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/122580)*

This is pretty cooked, but I think it works.

compiler-builtins has a long-standing problem that at link time, its rlib cannot contain any calls to `core`. And yet, in codegen we _love_ inserting calls to symbols in `core`, generally from various panic entrypoints.

I intend this PR to attack that problem as completely as possible. When we generate a function call, we now check if we are generating a function call from `compiler_builtins` and whether the callee is a function which was not lowered in the current crate, meaning we will have to link to it.

If those conditions are met, actually generating the call is asking for a linker error. So we don't. If the callee diverges, we lower to an abort with the same behavior as `core::intrinsics::abort`. If the callee does not diverge, we produce an error. This means that compiler-builtins can contain panics, but they'll SIGILL instead of panicking. I made non-diverging calls a compile error because I'm guessing that they'd mostly get into compiler-builtins by someone making a mistake while working on the crate, and compile errors are better than linker errors. We could turn such calls into aborts as well if that's preferred.

### CODE REVIEW

File: compiler/rustc_codegen_ssa/src/mir/block.rs [hunk 1]
```diff
@@ -157,8 +160,31 @@ impl<'a, 'tcx> TerminatorCodegenHelper<'tcx> {
...
+                    info!(
+                        "compiler_builtins call to diverging function {:?} replaced with abort",
+                        instance.def_id()
+                    );
+                    bx.abort();
+                    bx.unreachable();
+                    return MergingSucc::False;
+                }
+            }
+        }
```
Comment (bjorn3) [inline compiler/rustc_codegen_ssa/src/mir/block.rs:186]: cg_clif will need this change too.

Comment (saethlin) [reply compiler/rustc_codegen_ssa/src/mir/block.rs:186]: If I apply this change to cranelift, how would I verify that it works?

Comment (bjorn3) [reply compiler/rustc_codegen_ssa/src/mir/block.rs:186]: https://github.com/rust-lang/rustc_codegen_cranelift/#building-and-testing-with-changes-in-rustc-code has instructions.

Comment (bjorn3) [reply compiler/rustc_codegen_ssa/src/mir/block.rs:186]: And the new code would likely be somewhere around https://github.com/rust-lang/rustc_codegen_cranelift/blob/4cf4ffc6ba514f171b3f52d1c731063e4fc45be3/src/abi/mod.rs#L374

**Comment by Amanieu** [comment]
I like this approach since it removes the need for existing hacks like disabling debug assertions and disabling overflow checks.

**Comment by saethlin** [comment]
That's also why I like it! For the record, @dpaoliello deserves some credit for saying this:

> If we can't rely on libcore, would it be possible to implement a "mini panic" in compiler-builtins instead?

Which taken literally is not possible, but a very good line of thinking.

**Comment by pnkfelix** [comment]
Hmm.

Is there any way to *test* the behavior being introduced here as part of our CI test suite?

It seems like testing it would require either:

1. substituting in a custom "buggy" version of compiler_builtins solely for testing purposes. Which does not sound feasible, or at least not a good use of effort if its solely to support testing this PR. Or,

2. *maybe* there is a way to make a custom variant of `libcore` and link to that, and have that custom `libcore` provide items that would normally return true for `should_codegen_locally` but now they return `false` for this hypothetical custom `libcore`..., or

3. add a new item to compiler_builtins that is itself designed to exercise the behavior being added here, but otherwise serves no purpose.
 
... so after some reflection, I am guessing that Options 1 and 2 are impossible or non-desirable; option 3 might be doable but isn't really attractive from an overall development trajectory standpoint.

And instead, we just say "we do not intend to regression test this PR as part of our test suite. One can test it in a local fashion by making local changes to compiler-builtins that expose the behavior in question."

Does that all sound right?

**Comment by pnkfelix** [comment]
This looks fine to me as it stands, assuming that I'm correct about it not being reasonable to try to test it in CI.

But I also am inferring that it should not land without the [corresponding changes to cg_clif](https://github.com/rust-lang/rust/pull/122580/files#r1527128138)

@rustbot author

**Comment by bjorn3** [comment]
> But I also am inferring that it should not land without the [corresponding changes to cg_clif](https://github.com/rust-lang/rust/pull/122580/files#r1527128138)

This has been added to the PR since I made that comment.

**Comment by pnkfelix** [comment]
> This has been added to the PR since I made that comment.

Oh, whoops! Thanks!

**Comment by saethlin** [comment]
I agree that we should have some tests for this. I'll open an issue with some ideas.

**Comment by Amanieu** [comment]
This is a false positive: the `sqrtss` function here is defined as:
```rust
extern "C" {
    #[link_name = "llvm.x86.sse.sqrt.ss"]
    fn sqrtss(a: __m128) -> __m128;
}
```

Since this is an LLVM intrinsic and not a real function, this should be allowed.

**Comment by saethlin** [comment]
I figured out how to write a test. I think.

File: tests/run-make/compiler-builtins/rmake.rs [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+//! resolves symbols first. So the portable and safe approach is to forbid such a linkage
+//! requirement entirely.
+//!
+//! In addition, whether compiler_builtins requires linkage against core can depend on optimization
+//! settings. Turning off optimizations and enabling debug assertions tends to produce the most
+//! dependence on core that is possible, so that is the configuration we test here.
+
+extern crate run_make_support;
+
+use run_make_support::goblin::archive::Archive;
```
Comment (bjorn3) [inline tests/run-make/compiler-builtins/rmake.rs:13]: Object supports reading archives too.

**Comment by saethlin** [comment]
Oh 🤦 there's no `cargo` on the PATH.

**Comment by saethlin** [comment]
@pnkfelix CI for the test finally passes, and locally I made sure it passes the test-various job. I'd appreciate your review of the new commit.

**Comment by pnkfelix** [comment]
Awesome job, I was expecting this to end up landing without a test but you really came through @saethlin !

@bors r+

File: tests/run-make/compiler-builtins/rmake.rs [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    undefined_relocations
+        .retain(|symbol| !defined_symbols.contains(symbol) && symbol.contains("core"));
+
+    if !undefined_relocations.is_empty() {
+        panic!(
+            "compiler_builtins must not link against core, but it does. \n\
+            These symbols may be undefined in a debug build of compiler_builtins:\n\
+            {:?}",
+            undefined_relocations
+        );
```
Comment (RalfJung) [inline tests/run-make/compiler-builtins/rmake.rs:140]: Why is this extra check needed, given the check during codegen? How can it even fail?

**Comment by Amanieu** [comment]
While trying to solve the CI issues in compiler_builtins, I noticed that `-C embed-bitcode=yes` causes references to panicking functions in core to be emitted.

That's not what's causing the currently issues in CI (x86_64-apple-darwin seems to still be generating references to panic), but it might be worth fixing.

File: tests/run-make/compiler-builtins/rmake.rs (hunk 3 above)
Comment (RalfJung) [reply tests/run-make/compiler-builtins/rmake.rs:140]: Okay so this kind of sanity-checks the codegen-time checks? Makes sense, thanks. (That may be worth a comment in the code though so it's also clear to a future reader.)

**Comment by saethlin** [comment]
> While trying to solve the CI issues in compiler_builtins, I noticed that `-C embed-bitcode=yes` causes references to panicking functions in core to be emitted.

I just noticed this as well. I'm starting to look into it.

**Comment by saethlin** [comment]
Ah no what I was looking at is another case of https://github.com/rust-lang/rust/pull/118770, at least this time it's just me causing trouble for me. I'll extract a reproducer from the compiler-builtins CI.

**Comment by petrochenkov** [comment]
@saethlin
The added run-make test fails with this message for me:
```
   Compiling core v0.0.0 (E:\msys64\home\we\rust\build\x86_64-pc-windows-gnu\stage1\lib\rustlib\src\rust\library\core)
   Compiling compiler_builtins v0.1.109
error: couldn't create a temp dir: Access is denied. (os error 5) at path "C:\\Windows\\rustcKMWsUo"

error: could not compile `compiler_builtins` (build script) due to 1 previous error
warning: build failed, waiting for other jobs to finish...
error: couldn't create a temp dir: Access is denied. (os error 5) at path "C:\\Windows\\rustclRyY6G"

error: could not compile `core` (lib) due to 1 previous error
thread 'main' panicked at E:\msys64\home\we\rust\tests\run-make\compiler-builtins\rmake.rs:71:5:
assertion failed: status.success()
stack backtrace:
   0: rust_begin_unwind
             at E:\msys64\home\we\rust\library\std\src\panicking.rs:652:5
   1: core::panicking::panic_fmt
             at E:\msys64\home\we\rust\library\core\src\panicking.rs:72:14
   2: core::panicking::panic
             at E:\msys64\home\we\rust\library\core\src\panicking.rs:146:5
   3: rmake::main
   4: core::ops::function::FnOnce::call_once
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

Looks like the test suite uses a wrong temporary directory on Windows in some cases.
Or MSYS2 is to blame, or the fact that my Rust repo is on disk `E:`, I haven't investigated yet.
cc @ChrisDenton

**Comment by jieyouxu** [comment]
> Looks like the test suite uses a wrong temporary directory on Windows in some cases. Or MSYS2 is to blame, or the fact that my Rust repo is on disk `E:`, I haven't investigated yet.

@petrochenkov Judging from the error message, it's emitted from codegen: https://github.com/rust-lang/rust/blob/22f5bdc42b06e0719eda6e3e23f28fe5493b2305/compiler/rustc_codegen_ssa/src/back/link.rs#L99-L102

Where somehow [`env::temp_dir`](https://doc.rust-lang.org/std/env/fn.temp_dir.html#) looks messed up -- I think this might be a msys2 weird interaction with `$TMPDIR` problem -- because apparently compiletest uses and modifies `$TMPDIR` (even though that env var influences what `env::temp_dir` returns!)

**Comment by ChrisDenton** [comment]
Oh it's using `env_clear`, which is always risky on Windows because there's at least a few environment variables that are assumed to always be set.

In this case either `TMP` or `TEMP` needs to be set otherwise getting the temp dir will fallback to `USERPROFILE` and if that's not found either then the windows directory is used as a last resort. Which will be unwritable.

**Comment by ChrisDenton** [comment]
I think the helper lib could have a `env_reset` that preserves some necessary environment variables rather than fully clearing everything.

**Comment by jieyouxu** [comment]
> I think the helper lib could have a `env_reset` that preserves some necessary environment variables rather than fully clearing everything.

I'll work on a patch for this

**Comment by jieyouxu** [comment]
> I think the helper lib could have a `env_reset` that preserves some necessary environment variables rather than fully clearing everything.

Opened https://github.com/rust-lang/rust/pull/125426, for now I only preserve `TMP` and `TEMP` on Windows, `TMPDIR` on non-Windows. Let me know if you know of any more "necessary" env vars.

**Comment by RalfJung** [comment]
> In this case either TMP or TEMP needs to be set otherwise getting the temp dir will fallback to USERPROFILE and if that's not found either then the windows directory is used as a last resort. Which will be unwritable.

Isn't that a terrible fallback? If that's happening in our implementation, it seems better to error out thatn provide an unwriteable directory.

**Comment by ChrisDenton** [comment]
It's not our fallback. See https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-gettemppath2w

**Comment by RalfJung** [comment]
Ah, unfortunate. I wonder if that's worth working around on our side. But `temp_dir` doesn't even return a `Result` so there's no good way we can indicate failure either...

**Comment by jieyouxu** [comment]
> Ah, unfortunate. I wonder if that's worth working around on our side. But `temp_dir` doesn't even return a `Result` so there's no good way we can indicate failure either...

We could still describe this in the docs, because it is very surprising...

**Comment by eddyb** [comment]
Just to not lose track of this before anyone opens an issue/fix (cc @saethlin @oli-obk), I stumbled upon an interaction where unoptimized `compiler_builtins` UB checks can get reused downstream (in very niche situations, mind you, I barely found any examples to demo this), replacing the panic w/ an `abort`:
```console
$ echo -e 'fn main(){unsafe{let _=[0u8].get_unchecked_mut(2..);}}' > main.rs && RUSTFLAGS='-Csymbol-mangling-version=v0' cargo +nightly-2025-06-23 -Zscript main.rs
warning: `package.edition` is unspecified, defaulting to `2024`
   Compiling main v0.0.0 (/tmp/tmp.a7HqAyaIKI/main.rs)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.20s
     Running `/home/eddy/.cargo/target/69/34d2a89093b915/debug/main`

thread 'main' panicked at /home/eddy/.rustup/toolchains/nightly-2025-06-23-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/slice/index.rs:554:44:
unsafe precondition(s) violated: slice::get_unchecked_mut requires that the range is within the slice

This indicates a bug in the program. This Undefined Behavior check is optional, and cannot be relied on for safety.
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread caused non-unwinding panic. aborting.
fish: Job 1, 'RUSTFLAGS='-Csymbol-mangling-ve…' terminated by signal SIGABRT (Abort)
$ nm ~/.cargo/target/69/34d2a89093b915/debug/main | rg '4main$' | rustfilt | rg precondition
0000000000015000 t <core::ops::range::Range<usize> as core::slice::index::SliceIndex<[_]>>::get_unchecked_mut::precondition_check
$ echo -e 'fn main(){unsafe{let _=[0u8].get_unchecked_mut(2..);}}' > main.rs && RUSTFLAGS='-Csymbol-mangling-version=v0' cargo +nightly-2025-06-23 -Zscript -Zbuild-std main.rs
warning: `package.edition` is unspecified, defaulting to `2024`
   Compiling main v0.0.0 (/tmp/tmp.a7HqAyaIKI/main.rs)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.23s
     Running `/home/eddy/.cargo/target/69/34d2a89093b915/debug/main`
fish: Job 1, 'RUSTFLAGS='-Csymbol-mangling-ve…' terminated by signal SIGILL (Illegal instruction)
$ nm ~/.cargo/target/69/34d2a89093b915/debug/main | rg '4main$' | rustfilt | rg precondition
```
(AIUI, this requires `-Zshare-generics`, and therefore `-Zbuild-std`, so it seems pretty unlikely anyone would hit this, and my initial suspect of `u32::unchecked_add` isn't even affected because only a generic `fn` like `Range<u32>::next` actually gets reused, and that should never fail the UB check)

(**EDIT**: removed a mention of a different PR that may not be as relevant as I thought)

**Comment by saethlin** [comment]
I've been thinking about completely special-casing the checks for compiler-builtins, because https://github.com/rust-lang/rust/pull/134938 is incompatible with this PR, in their current forms. Not sure how to split the difference but yeah I think it's clear there is some kind of design flaw here.

**Comment by RalfJung** [comment]
> If those conditions are met, actually generating the call is asking for a linker error. So we don't. If the callee diverges, we lower to an abort with the same behavior as core::intrinsics::abort. 

I completely missed this. What is going on here? We have the backend silently replace a call of a function we don't know anything about except that it will not return, with an `abort`? This can completely change the behavior of the program! I cannot see how we could argue that this is correct compilation.

File: compiler/rustc_codegen_ssa/src/mir/block.rs [hunk 4]
```diff
@@ -1664,11 +1693,15 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
 
         self.set_debug_loc(&mut bx, mir::SourceInfo::outermost(self.mir.span));
 
-        let (fn_abi, fn_ptr) = common::build_langcall(&bx, None, reason.lang_item());
-        let fn_ty = bx.fn_decl_backend_type(fn_abi);
+        let (fn_abi, fn_ptr, instance) = common::build_langcall(&bx, None, reason.lang_item());
+        if is_call_from_compiler_builtins_to_upstream_monomorphization(bx.tcx(), instance) {
+            bx.abort();
+        } else {
```
Comment (RalfJung) [inline compiler/rustc_codegen_ssa/src/mir/block.rs:1699]: And then there's another "replace call by abort" thing here that doesn't even have a comment explaining why we are silently patching the code while compiling it.

**Pull request merged**